File: LeetCode_Algorithm/100(tree)/100.js
function TreeNode(val, left, right) {
  this.val = (val===undefined ? 0 : val)
  this.left = (left===undefined ? null : left)
  this.right = (right===undefined ? null : right)
}
function makeTree(lst, idx) {
  if (idx > lst.length-1) {
    return null
  }
  let root = new TreeNode(lst[idx])
  root.left = makeTree(lst, 2*idx+1)
  root.right = makeTree(lst, 2*idx+2)
  return root
}

var isSameTree = function(p, q) {
  let pNode = makeTree(p,0)
  let qNode = makeTree(q,0)
  // O(n) O(1) 87ms 48%
  lst = []; lst.push([pNode,qNode])
  while (lst.length > 0) {
    temp = lst.shift()
    pNode = temp[0]; qNode = temp[1]
    // console.log(pNode, qNode)
    if (pNode == null && qNode == null) {
      continue
    }
    if (!(pNode) || !(qNode)) {
      // console.log(pNode, qNode)
      return false
    }
    if (pNode.val != qNode.val) {
      // console.log(pNode, qNode)
      return false
    }
    lst.push([pNode.left, qNode.left])
    lst.push([pNode.right, qNode.right])
  } 
  return true
};

console.log(isSameTree([1,2,3], [1,2,3]))

File: LeetCode_Algorithm/100(tree)/100.py
from sys import setdlopenflags
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        p_root = self.makeTree(p, 0)
        q_root = self.makeTree(q, 0)
        return self.solve(p_root, q_root)

    # sol1 O(n) O(n) 68ms 6%
    def solve(self, p, q):
        # 둘 다 None
        if not p and not q:
            return True
        if not p or not q:
            return False
        if p.val != q.val:
            return False
        return self.solve(p.left, q.left) and self.solve(p.right, q.right)


    def makeTree(self, lst, idx):
        if idx >= len(lst):
            return None
        root = TreeNode(lst[idx])
        root.left = self.makeTree(lst, 2*idx+1)
        root.right = self.makeTree(lst, 2*idx+2)
        return root

print(Solution().isSameTree([1,2,3], [1,2,3]))